content/en/docs/v3.6/learning/api_guarantees.md

```diff
@@ -85,18 +70,36 @@ most current value. Without linearizability guarantee, the returned value,
 current at *t2* when the read began, might be "stale" by *t3* because a
 concurrent write might happen between *t2* and *t3*.
 
-etcd does not ensure linearizability for watch operations. Users are expected
-to verify the revision of watch responses to ensure correct ordering.
-
 etcd ensures linearizability for all other operations by default.
 Linearizability comes with a cost, however, because linearized requests must go
 through the Raft consensus process. To obtain lower latencies and higher
 throughput for read requests, clients can configure a request’s consistency mode
 to `serializable`, which may access stale data with respect to quorum, but
 removes the performance penalty of linearized accesses' reliance on live consensus.
 
+## Watch Service
+
+Watches make guarantees about events:
+* Ordered - events are ordered by revision.
+  An event will never appear on a watch if it precedes an event in time that
+  has already been posted.
+* Unique - an event will never appear on a watch twice.
+* Reliable - a sequence of events will never drop any subsequence of events
+  within a compaction window. Watch may be terminated if it's no longer in
+  history window.
+* Atomic - a list of events is guaranteed to encompass complete revisions.
+  Updates in the same revision over multiple keys will not be split over several
+  lists of events.
+* Resumable - A broken watch can be resumed by establishing a new watch starting
+  after the last revision received in a watch event before the break, so long as
+  the revision is in the history window.
+* Bookmarkable - Progress notification events guarantee that all events up to a
+  revision have been already delivered.
```

Comment:
I don't think `Resumable` and `Bookmarkable` are at the same level as other guarantees; instead, they can be inferred from other meta items. Suggest to either remove them or change it to something like below,
```suggestion

From the above guarantees, the following two items can be deduced,
* Resumable - A broken watch can be resumed by establishing a new watch starting
  after the last revision received in a watch event before the break, so long as
  the revision is in the history window.
* Bookmarkable - Progress notification events guarantee that all events up to a
  revision have been already delivered.
```

Reply:
Please read the discussion on the document https://docs.google.com/document/d/1NUZDiJeiIH5vo_FMaTWf0JtrQKCx0kpEaIIuPoj9P6A/edit?disco=AAAAv5-fWt0

Separation of Resumable from Bookmarkable was proposed by @jpbetz. Paraphrasing:
There are many streaming services that do not provide resumable guarantee, so the fact that etcd does is pretty important. For example zookeeper does not provide resumable. https://zookeeper.apache.org/doc/current/zookeeperProgrammers.html#sc_WatchSemantics

Reply:
Let's see other maintainers' opinion. cc @mitake @ptabor @spzala

Reply:
Linked document was already shared with all [maintainers over two weeks ago](https://groups.google.com/g/etcd-maintainers/c/wuUr5TZaw_E), so I don't think we will get more. I encourage everyone to join the discussion on the document as this represents agreement between etcd and Kubernetes communities.

Reply:
Sorry for any confusion. I was saying that instead of removing the proposed changes we keep it but I was thinking in the same line of write up as @ahrtr suggesting. As I said, I don't have strong opinion. I would vote to keep them in whatever we decide at the end. Thanks!

Reply:
As mentioned above, there is no implication from {Ordered, Unique, Reliable, Atomic} to {Resumable, Bookmarkable}.

Reply:
> Linked document was already shared with all maintainers over two weeks ago, so I don't think we will get more.

This PR is changing etcd's doc, I am confused why don't we need to get other maintainers' comments? No matter what other doc is, it doesn't mean that we need to update etcd following exactly the same way.

> As mentioned above, there is no implication from {Ordered, Unique, Reliable, Atomic} to {Resumable, Bookmarkable} so the write up is incorrect.

If a watch somehow is broken, the users just need to kick off a new watch starting from the last revision received in a watch event before the break, then etcd still guarantees `{Ordered, Unique, Reliable, Atomic}`, doesn't it mean what you called "Resumable"? 

FYI.  when there is temporary network disconnection between clientv3 and etcdserver, the etcd clientv3 automatically recovers or resumes the watch. https://github.com/etcd-io/etcd/pull/15248#discussion_r1100631660

The same thing to `Bookmarkable`. Thanks to that etcd guarantees {Ordered, Unique, Reliable, Atomic}, then when a client receives a revision, then all events up to the revision must have been already delivered, as long as there is compaction in between.

Please let me know if I misunderstood your point.

Reply:
> > Linked document was already shared with all maintainers over two weeks ago, so I don't think we will get more.
> 
> This PR is changing etcd's doc, I am confused why don't we need to get other maintainers' comments? No matter what other doc is, it doesn't mean that we need to update etcd following exactly the same way.

This is not what I mean. I said that document was already sent to all maintainers two weeks ago so I would expect they would have comment on the document if they had something to day about the contract itself.

> > As mentioned above, there is no implication from {Ordered, Unique, Reliable, Atomic} to {Resumable, Bookmarkable} so the write up is incorrect.
> 
> If a watch somehow is broken, the users just need to kick off a new watch starting from the last revision received in a watch event before the break, then etcd still guarantees `{Ordered, Unique, Reliable, Atomic}`, doesn't it mean what you called "Resumable"?

What I'm meant is that implication is the other direction. `Resumable` requires `{Ordered, Unique, Reliable, Atomic}`, but fact that database has `{Ordered, Unique, Reliable, Atomic}`, doesn't mean it needs to provide `Resumable`. By fact that etcd has compaction `Resumable` is already limited to available history window. You could technically implement database that never compacted, and database that immediately compacted after a request. So the fact that etcd provides `Resumable` and provide it's limitation is important for users that want to build applications on etcd like Kubernetes.

> FYI. when there is temporary network disconnection between clientv3 and etcdserver, the etcd clientv3 automatically recovers or resumes the watch. [etcd-io/etcd#15248 (comment)](https://github.com/etcd-io/etcd/pull/15248#discussion_r1100631660)

Sure, however etcd client resume has its own limits. Kubernetes implements their own resume mechanism.

> The same thing to `Bookmarkable`. Thanks to that etcd guarantees {Ordered, Unique, Reliable, Atomic}, then when a client receives a revision, then all events up to the revision must have been already delivered, as long as there is NO compaction in between.
> 
> Please let me know if I misunderstood your point.

Bookmarkable is required as progress notification is a special type of watch response. Watch response will include events with their mod revision and the last committed revision. Normally mod revisions are unrelated from committed revision, however in progress notification there are no events and last committed revision needs to be sent after all events with mod revision below it have been delivered.

This property was important in finding https://github.com/etcd-io/etcd/pull/15248

Reply:
OK, makes some sense to me.

Reply:
thx